proposals/NNNN-library-evolution.md

```diff
@@ -0,0 +1,370 @@
+
+Some ideas for how to prevent library authors from breaking the rules accidentally are discussed in "Compatibility checking" under "Future directions".
+
+
+## Future directions
+
+### Compatibility checking
+
+Of course, the compiler can't stop a library author from modifying the fields of a fixed-contents struct, even though that will break binary compatibility. We already have two ideas on how we could catch mistakes of this nature:
+
+- A checker that can compare APIs across library versions, using swiftmodule files or similar.
+
+- Encoding the layout of a type in a symbol name. Clients could link against this symbol so that they'd fail to launch if it changes, but even without that an automated system could check the list of exported symbols to make sure nothing was removed.
+
+
+### Allowing observing accessors
+
+The limitation on observing accessors for stored instance properties is an artificial one; clients could certainly access the field directly for "read" accesses but go through a setter method for "write" accesses. However, the expected use for `@frozen` is for structs that need to be as efficient as C structs, and that includes being able to do direct stores into public instance properties. The effect of observing accessors can be emulated in a fixed-contents struct with a private stored property and a public computed property.
+
+If we wanted to add this later, the way we would probably implement it is by saying that accessors for stored instance properties of `@frozen` structs must be marked `@inlinable`, with all of the implications that has for being able to make changes in the future.
+
+
+## Alternatives considered
+
+### Annotation syntax
+
+#### Naming
+
+`@frozen` comes from [SE-0192][], where it was originally used as a term for for enums that will not gain new cases.
+
+Other than `@frozen`, most names have revolved around the notion of "fixed" somehow: `@fixedContents`, `@fixedLayout`, or even just `@fixed`. We ultimately chose `@frozen` as there is a strong similarity between this feature for both enums and structs.
+
+`final` was suggested at one point, but the meaning is a little far from that of `final` on classes.
+
+#### Modifier or attribute?
+
+This proposal suggests a new *attribute* for structs, `@fixedContents`; it could also be a modifier `fixedContents` (or `fixed`), implemented as a context-sensitive keyword. Because this annotation only affects the struct's *implementation* rather than anything about its use, an attribute seemed more appropriate.
```

Comment:
```suggestion
This proposal suggests a new *attribute* for structs, `@frozen`; it could also be a modifier `frozen`, implemented as a context-sensitive keyword. Because this annotation only affects the struct's *implementation* rather than anything about its use, an attribute seemed more appropriate.
```

Reply:
Ah, needed to be generilized for enums too.